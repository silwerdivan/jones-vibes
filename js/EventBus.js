// A simple pub/sub module for decoupled communication.
const EventBus = {
  events: {},

  /**
   * Subscribe to an event.
   * @param {string} eventName The name of the event (e.g., 'stateChanged').
   * @param {function} callback The function to execute when the event is published.
   */
  subscribe(eventName, callback) {
    if (!this.events[eventName]) {
      this.events[eventName] = [];
    }
    this.events[eventName].push(callback);
  },

  /**
   * Publish an event.
   * @param {string} eventName The name of the event to publish.
   * @param {*} data The data to pass to all subscribed callbacks.
   */
  publish(eventName, data) {
    if (this.events[eventName]) {
      this.events[eventName].forEach(callback => {
        // Pass the data to the callback.
        callback(data);
      });
    }
  }
};

export default EventBus;
