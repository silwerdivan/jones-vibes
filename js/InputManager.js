class InputManager {
  /**
   * @param {GameController} gameController An instance of the GameController.
   */
  constructor(gameController) {
    this.gameController = gameController;
  }

  initialize() {
    const gameContainer = document.querySelector('.game-container');
    const eventType = 'ontouchstart' in window ? 'touchstart' : 'click';

    // Use a single listener on the parent container for efficiency.
    gameContainer.addEventListener(eventType, (event) => {
      // Find the closest parent element (or the element itself) with a data-action attribute.
      const target = event.target.closest('[data-action]');
      
      if (!target) {
        return; // Exit if the click was not on an actionable element.
      }

      const action = target.dataset.action;

      // Check if a method with this name exists on the controller.
      if (typeof this.gameController[action] === 'function') {
        // Call the corresponding method on the game controller.
        this.gameController[action]();
      } else {
        console.warn(`Action "${action}" is not defined on GameController.`);
      }
    });
  }
}

export default InputManager;
