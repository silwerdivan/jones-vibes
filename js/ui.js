// js/ui.js

export function render(gameState) {
    if (!gameState || !gameState.players) {
        console.error("Invalid gameState object provided to render function.");
        return;
    }

    // Update player panels
    gameState.players.forEach((player, index) => {
        const playerPanelId = `#p${index + 1}-`;
        document.querySelector(`${playerPanelId}cash`).textContent = player.cash;
        document.querySelector(`${playerPanelId}savings`).textContent = player.savings;
        document.querySelector(`${playerPanelId}loan`).textContent = player.loan;
        document.querySelector(`${playerPanelId}happiness`).textContent = player.happiness;
        document.querySelector(`${playerPanelId}education`).textContent = player.educationLevel;
        document.querySelector(`${playerPanelId}career`).textContent = player.careerLevel;
        document.querySelector(`${playerPanelId}car`).textContent = player.hasCar ? 'Yes' : 'No';
        document.querySelector(`${playerPanelId}time`).textContent = player.time;

        // Highlight current player
        const panel = document.querySelector(`#player${index + 1}-status-panel`);
        if (index === gameState.currentPlayerIndex) {
            panel.classList.add('current-player');
        } else {
            panel.classList.remove('current-player');
        }
    });

    // Update location info
    document.querySelector('#current-location').textContent = gameState.getCurrentPlayer().location;
    document.querySelector('#game-turn').textContent = gameState.turn;

    // Show/hide action buttons based on location
    const currentPlayer = gameState.getCurrentPlayer();
    const locationSpecificButtons = document.querySelectorAll('.location-specific');
    locationSpecificButtons.forEach(button => button.classList.add('hidden')); // Hide all first

    switch (currentPlayer.location) {
        case 'Employment Agency':
            document.querySelector('#btn-work-shift').classList.remove('hidden');
            break;
        case 'Community College':
            document.querySelector('#btn-take-course').classList.remove('hidden');
            break;
        case 'Shopping Mall':
            document.querySelector('#btn-buy-item').classList.remove('hidden');
            break;
        case 'Bank':
            document.querySelector('#btn-deposit').classList.remove('hidden');
            document.querySelector('#btn-withdraw').classList.remove('hidden');
            document.querySelector('#btn-take-loan').classList.remove('hidden');
            document.querySelector('#btn-repay-loan').classList.remove('hidden');
            document.querySelector('#btn-buy-car').classList.remove('hidden');
            break;
        default:
            // No specific buttons for other locations like 'Home' or 'Start'
            break;
    }
    document.querySelector('#btn-travel').classList.remove('hidden'); // Travel is always an option
}