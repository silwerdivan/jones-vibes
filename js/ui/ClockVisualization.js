class ClockVisualization {
    constructor(containerId, options = {}) {
        this.container = document.getElementById(containerId);
        if (!this.container) {
            throw new Error(`Container with id "${containerId}" not found`);
        }

        // Default options
        this.options = {
            size: options.size || 60,
            strokeWidth: options.strokeWidth || 8,
            backgroundColor: options.backgroundColor || 'rgba(0, 255, 255, 0.2)',
            foregroundColor: options.foregroundColor || '#00FFFF',
            textColor: options.textColor || '#FFFFFF',
            fontSize: options.fontSize || '14px',
            fontFamily: options.fontFamily || 'VT323, monospace',
            showNumeric: options.showNumeric !== undefined ? options.showNumeric : true,
            animationDuration: options.animationDuration || 500,
            ...options
        };

        this.currentHours = 24;
        this.init();
    }

    init() {
        // Create the SVG element for the clock
        this.svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
        this.svg.setAttribute('width', this.options.size);
        this.svg.setAttribute('height', this.options.size);
        this.svg.setAttribute('viewBox', `0 0 ${this.options.size} ${this.options.size}`);
        this.svg.style.display = 'block';

        // Create background circle
        this.backgroundCircle = document.createElementNS('http://www.w3.org/2000/svg', 'circle');
        this.backgroundCircle.setAttribute('cx', this.options.size / 2);
        this.backgroundCircle.setAttribute('cy', this.options.size / 2);
        this.backgroundCircle.setAttribute('r', (this.options.size / 2) - (this.options.strokeWidth / 2));
        this.backgroundCircle.setAttribute('fill', 'none');
        this.backgroundCircle.setAttribute('stroke', this.options.backgroundColor);
        this.backgroundCircle.setAttribute('stroke-width', this.options.strokeWidth);
        this.svg.appendChild(this.backgroundCircle);

        // Create foreground circle (the pie chart)
        this.foregroundCircle = document.createElementNS('http://www.w3.org/2000/svg', 'circle');
        this.foregroundCircle.setAttribute('cx', this.options.size / 2);
        this.foregroundCircle.setAttribute('cy', this.options.size / 2);
        this.foregroundCircle.setAttribute('r', (this.options.size / 2) - (this.options.strokeWidth / 2));
        this.foregroundCircle.setAttribute('fill', 'none');
        this.foregroundCircle.setAttribute('stroke', this.options.foregroundColor);
        this.foregroundCircle.setAttribute('stroke-width', this.options.strokeWidth);
        this.foregroundCircle.setAttribute('stroke-dasharray', '0 1000');
        this.foregroundCircle.setAttribute('transform', `rotate(-90 ${this.options.size / 2} ${this.options.size / 2})`);
        this.foregroundCircle.style.transition = `stroke-dasharray ${this.options.animationDuration}ms ease-in-out`;
        this.svg.appendChild(this.foregroundCircle);

        // Create text element for numeric time display
        if (this.options.showNumeric) {
            this.textElement = document.createElementNS('http://www.w3.org/2000/svg', 'text');
            this.textElement.setAttribute('x', this.options.size / 2);
            this.textElement.setAttribute('y', this.options.size / 2);
            this.textElement.setAttribute('text-anchor', 'middle');
            this.textElement.setAttribute('dominant-baseline', 'middle');
            this.textElement.setAttribute('fill', this.options.textColor);
            this.textElement.setAttribute('font-size', this.options.fontSize);
            this.textElement.setAttribute('font-family', this.options.fontFamily);
            this.textElement.textContent = '24h';
            this.svg.appendChild(this.textElement);
        }

        // Clear container and append the SVG
        this.container.innerHTML = '';
        this.container.appendChild(this.svg);

        // Set initial state
        this.updateTime(24);
    }

    updateTime(hours) {
        // Ensure hours is within valid range
        hours = Math.max(0, Math.min(24, hours));
        this.currentHours = hours;

        // Calculate the circumference
        const radius = (this.options.size / 2) - (this.options.strokeWidth / 2);
        const circumference = 2 * Math.PI * radius;

        // Calculate the filled portion (24 hours = empty, 0 hours = full)
        const filledPercentage = (24 - hours) / 24;
        const dashLength = circumference * filledPercentage;
        const gapLength = circumference - dashLength;

        // Update the stroke-dasharray to create the pie chart effect
        this.foregroundCircle.setAttribute('stroke-dasharray', `${dashLength} ${gapLength}`);

        // Update the text if enabled
        if (this.options.showNumeric && this.textElement) {
            this.textElement.textContent = `${hours}h`;
        }

        // Calculate color intensity based on time remaining (darker as time depletes)
        const intensityPercentage = 1 - (hours / 24); // 0 when full time, 1 when no time
        const darkerColor = this.getDarkerColor(this.options.foregroundColor, intensityPercentage);
        
        // Apply the darker color to the foreground circle
        this.foregroundCircle.setAttribute('stroke', darkerColor);
        
        // Also update text color to match for consistency
        if (this.options.showNumeric && this.textElement) {
            this.textElement.setAttribute('fill', darkerColor);
        }
    }

    // Helper method to calculate a darker version of a color
    getDarkerColor(color, intensity) {
        // If intensity is 0, return the original color
        if (intensity === 0) return color;
        
        // Parse the color to RGB values
        let r, g, b;
        
        if (color.startsWith('#')) {
            // Hex color format
            const hex = color.substring(1);
            r = parseInt(hex.substring(0, 2), 16);
            g = parseInt(hex.substring(2, 4), 16);
            b = parseInt(hex.substring(4, 6), 16);
        } else if (color.startsWith('rgb')) {
            // RGB color format
            const matches = color.match(/\d+/g);
            r = parseInt(matches[0]);
            g = parseInt(matches[1]);
            b = parseInt(matches[2]);
        } else {
            // Default to cyan for named colors
            return color;
        }
        
        // Calculate darker values by reducing each component
        const darknessFactor = 1 - (intensity * 0.7); // Max 70% darker
        r = Math.floor(r * darknessFactor);
        g = Math.floor(g * darknessFactor);
        b = Math.floor(b * darknessFactor);
        
        // Convert back to hex
        return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
    }

    // Method to get current hours
    getCurrentHours() {
        return this.currentHours;
    }

    // Method to destroy the component
    destroy() {
        if (this.container && this.svg) {
            this.container.removeChild(this.svg);
        }
    }
}

export default ClockVisualization;