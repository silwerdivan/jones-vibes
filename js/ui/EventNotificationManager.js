import EventBus from '../EventBus.js';

class EventNotificationManager {
  constructor() {
    this.currentEvent = null;
    this.displayTimeout = null;
    this.unreadCount = 0;
    this.displayDuration = 2000; // 2 seconds minimum display time
    
    // DOM elements
    this.eventStrip = null;
    this.eventText = null;
    this.logIcon = null;
    this.badge = null;
    
    this.init();
  }
  
  init() {
    this.createEventStrip();
    this.createLogIcon();
    
    // Subscribe to game events
    EventBus.subscribe('gameEvent', (event) => {
      this.addEvent(event);
    });
    
    // Subscribe to log opened event
    EventBus.subscribe('logOpened', () => {
      this.resetUnreadCount();
    });
    
    // Subscribe to add to log event
    EventBus.subscribe('addToLog', (event) => {
      // This event is handled by the regular log update in GameState
      // We don't need to do anything here as the event is already in the log
    });
  }
  
  createEventStrip() {
    // Create event notification strip
    this.eventStrip = document.createElement('div');
    this.eventStrip.id = 'event-notification-strip';
    this.eventStrip.className = 'event-notification-strip hidden';
    
    this.eventText = document.createElement('div');
    this.eventText.className = 'event-notification-text';
    
    this.eventStrip.appendChild(this.eventText);
    
    // Insert above the action buttons panel
    const actionsPanel = document.getElementById('game-controls');
    actionsPanel.parentNode.insertBefore(this.eventStrip, actionsPanel);
  }
  
  createLogIcon() {
    // Create log icon with badge
    const logIconContainer = document.createElement('div');
    logIconContainer.id = 'log-icon-container';
    logIconContainer.className = 'log-icon-container';
    
    this.logIcon = document.createElement('div');
    this.logIcon.id = 'log-icon';
    this.logIcon.className = 'log-icon';
    this.logIcon.innerHTML = 'ðŸ“‹';
    
    this.badge = document.createElement('div');
    this.badge.id = 'notification-badge';
    this.badge.className = 'notification-badge hidden';
    
    logIconContainer.appendChild(this.logIcon);
    logIconContainer.appendChild(this.badge);
    
    // Add click handler to open log
    logIconContainer.addEventListener('click', () => {
      EventBus.publish('logIconClicked');
    });
    
    // Position it in the event log header
    const eventLogHeader = document.querySelector('.event-log h3');
    if (eventLogHeader) {
      eventLogHeader.appendChild(logIconContainer);
    }
  }
  
  addEvent(event) {
    // Clear any existing timeout
    if (this.displayTimeout) {
      clearTimeout(this.displayTimeout);
      this.displayTimeout = null;
    }
    
    // If there's a current event, send it to log first
    if (this.currentEvent) {
      EventBus.publish('addToLog', this.currentEvent);
    }
    
    // Set the new event as current
    this.currentEvent = event;
    this.unreadCount++;
    this.updateBadge();
    
    // Display the new event immediately
    this.displayEvent(event);
    
    // Set timeout to send current event to log after display duration
    this.displayTimeout = setTimeout(() => {
      this.sendCurrentEventToLog();
    }, this.displayDuration);
  }
  
  sendCurrentEventToLog() {
    if (this.currentEvent) {
      this.animateToLog(() => {
        EventBus.publish('addToLog', this.currentEvent);
        this.currentEvent = null;
        this.displayTimeout = null;
      });
    }
  }
  
  displayEvent(event) {
    // Set event text and category
    this.eventText.textContent = typeof event === 'string' ? event : event.text;
    
    // Add category class if available
    if (event && event.category) {
      this.eventText.className = `event-notification-text log-${event.category}`;
    } else {
      this.eventText.className = 'event-notification-text';
    }
    
    // Show the strip with animation
    this.eventStrip.classList.remove('hidden');
    this.eventStrip.classList.add('show');
  }
  
  animateToLog(callback) {
    // Add shrinking animation
    this.eventStrip.classList.add('shrinking');
    
    // After animation completes, hide the strip
    setTimeout(() => {
      this.eventStrip.classList.remove('show', 'shrinking');
      this.eventStrip.classList.add('hidden');
      
      if (callback) callback();
    }, 500); // Animation duration
  }
  
  updateBadge() {
    if (this.unreadCount > 0) {
      this.badge.textContent = this.unreadCount > 99 ? '99+' : this.unreadCount;
      this.badge.classList.remove('hidden');
    } else {
      this.badge.classList.add('hidden');
    }
  }
  
  resetUnreadCount() {
    this.unreadCount = 0;
    this.updateBadge();
  }
}

export default EventNotificationManager;