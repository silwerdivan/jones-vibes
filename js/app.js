import GameState from './game/GameState.js';
import GameController from './game/GameController.js';
import GameView from './ui.js';
import InputManager from './InputManager.js';

// The main entry point for the application.
function main() {
  // GameState can be initialized first
  const gameState = new GameState(2, true); // 2 players, P2 is AI

  // The View needs the Controller to create choice callbacks, but the Controller
  // needs the View to show modals. We solve this by creating them and then linking them.
  let gameController; // Declare upfront

  // 1. Initialize the View, passing a reference to the future controller.
  const gameView = new GameView(gameController);

  // 2. Initialize the controller, giving it access to the state AND the view.
  gameController = new GameController(gameState, gameView);

  // 3. NOW, assign the fully-built controller to the view.
  gameView.gameController = gameController;

  // 4. Initialize the input manager, giving it access to the controller.
  const inputManager = new InputManager(gameController);
  inputManager.initialize();

  // 5. Perform the initial render.

}

// Start the game when the DOM is ready.
document.addEventListener('DOMContentLoaded', main);