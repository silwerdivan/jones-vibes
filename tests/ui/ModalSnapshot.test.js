import { showChoiceModal } from '../../js/ui.js';

describe('Modal UI Snapshot', () => {
    let originalAppendChild;
    let originalRemoveChild;

    beforeAll(() => {
        // Mock appendChild and removeChild to prevent actual DOM manipulation during tests
        originalAppendChild = document.body.appendChild;
        originalRemoveChild = document.body.removeChild;
        document.body.appendChild = jest.fn();
        document.body.removeChild = jest.fn();
    });

    afterAll(() => {
        // Restore original functions
        document.body.appendChild = originalAppendChild;
        document.body.removeChild = originalRemoveChild;
    });

    beforeEach(() => {
        // Clear any existing modal content before each test
        document.body.innerHTML = '';
        jest.clearAllMocks();
    });

    test('should render the choice modal correctly when shown', async () => {
        const title = 'Test Modal Title';
        const message = 'This is a test modal message.'; // showChoiceModal doesn't use message directly
        const choices = [
            'Option A', 
            'Option B',
        ];

        // showChoiceModal returns a Promise, so we need to await it or handle it.
        // For snapshot testing, we just need it to render, not resolve.
        const modalPromise = showChoiceModal(title, choices);

        // The modal content is appended to document.body, so we snapshot the body
        expect(document.body.innerHTML).toMatchSnapshot();

        // Clean up the modal after snapshotting
        const overlay = document.getElementById('choice-modal-overlay');
        if (overlay) {
            overlay.remove();
        }

        // To prevent unhandled promise rejection warning, we can catch the promise
        modalPromise.catch(() => {});
    });
});
