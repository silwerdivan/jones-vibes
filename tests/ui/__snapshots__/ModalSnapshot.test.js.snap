// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Modal UI Snapshot should render the choice modal correctly when shown 1`] = `""`;
